### DISCUSSION TITLE
GC9503V support

### DISCUSSION THREAD

**Discussion by moinmoin-sh** [lovyan03/LovyanGFX#371] [Device Configurations]
Is there a way to use the GC9503V with your lib? 
It's used in this panel WT32S3-86V driven by an ESP32S3.
Thanks.

**Comment by Adrianotiger** [comment]
The GC9XXX is already present in the repository:
https://github.com/lovyan03/LovyanGFX/blob/master/src/lgfx/v1/panel/Panel_GC9A01.hpp

Maybe it works with it. If you can find the datasheet of the GC9503V, you can just add the init commands in this file (I can't find any datasheet with the communication protocol of this chip).

**Reply by lovyan03** [reply]
Please refer to this file for the setting method.
Since `Panel_ST7701` is used, you need to change it to `Panel_GC9503` .
https://github.com/lovyan03/LovyanGFX/blob/master/src/lgfx_user/LGFX_ESP32S3_RGB_MakerfabsParallelTFTwithTouch40.h


https://github.com/lovyan03/LovyanGFX/blob/master/src/lgfx/v1/platforms/esp32s3/Panel_RGB.hpp#L104
Implementation is here.

Please note that it may not work due to insufficient verification.

**Comment by moinmoin-sh** [comment]
That sounds like a good starting point. Thanks for your feedback.
I've a panel on route and will share the progress (if any ..)

**Reply by strange-v** [reply]
Hi @moinmoin-sh, any progress with WT32S3-86V? I've just started digging in the same direction.

**Reply by moinmoin-sh** [reply]
Hi @strange-v , yes there is. There was more discussion here under other topics.
I was finally able to integrate the WT32S3-86V in this project: https://github.com/OXRS-IO/OXRS-IO-TouchPanel-ESP32-FW
The configuration file can be found here: https://github.com/OXRS-IO/OXRS-IO-TouchPanel-ESP32-FW/blob/main/src/panels/cfgWT32S3-86V.hpp
Hope that helps.

**Reply by strange-v** [reply]
Thank you, it'll help a lot.

**Reply by strange-v** [reply]
Hmm... I'm dealing with a slightly newer WT32S3-86S panel (ZX3D95CE01S-TR-4848) though it uses the same GC9503V IC.
After adjusting the pins according to the datasheet, I got a working backlight and touch panel, but nothing is shown on the screen.

@lovyan03, could you please suggest whether the library has something like debug mode, or maybe other recommendations on how to investigate the issue?

Here is the  current code: [esp32s3_GC9503_panel_test](https://github.com/strange-v/esp32s3_GC9503_panel_test)

**Reply by tobozo** [reply]
@strange-v please make the [repository](https://github.com/strange-v/esp32s3_GC9503_panel_test) public 😉

**Reply by strange-v** [reply]
Oops, done. Thanks for noticing it.

**Reply by moinmoin-sh** [reply]
I don't have the 86S version panel on hand, but started to prepare the integration for someone else. Testing is ongoing.

**Reply by strange-v** [reply]
Here is the screen code from the manufacturer: https://github.com/wireless-tag-com/ZX3D95CE01S-TR-4848/blob/master/main/screen.c
It looks quite similar to what I have.

It would be interesting to understand how to verify the correctness of:
- hsync_polarity
- vsync_polarity
- pclk_idle_high
- de_idle_high

@moinmoin-sh, I can test whether it runs when the integration is ready.

**Reply by moinmoin-sh** [reply]
I could provide a binary if you are interested?

**Reply by strange-v** [reply]
Why not, let's try.

**Comment by strange-v** [comment]
To test whether the manufacturer provided the correct pinout and code, I built their ESP-IDF [sample project](https://github.com/wireless-tag-com/ZX3D95CE01S-TR-4848), and it does work. At least something :)

**Reply by strange-v** [reply]
Arduino_GFX works with the same parameters. I may use it, but LovyanGFX always had a better performance.

**Comment by moinmoin-sh** [comment]
This is the binary I anounced. 
_removed because not working_
I recommend to erase the flash and use the esptool to flash the binary at address 0.
If everything goes well you should see the blue OXRS splash screen which confirms the screen is working.

**Reply by strange-v** [reply]
The behavior is identical to my test code - a blank screen after the start.
If I switch power on/off a few times, I may catch this:
![image](https://github.com/lovyan03/LovyanGFX/assets/3875529/0b4a5c7a-43c4-4e26-a72f-b46ec3f23fe9)

It looks like something is wrong with the timings.

**Reply by moinmoin-sh** [reply]
That's a shame, thanks for testing.

There is one major thing different in the 86S version. The panel reset ist taken from the VSYNC 

![grafik](https://github.com/lovyan03/LovyanGFX/assets/53935853/e8ef2609-4c73-4bf4-8677-c525575a1b69)
![grafik](https://github.com/lovyan03/LovyanGFX/assets/53935853/f6e71898-03b5-47ae-b66d-41eb382a60ed)

This means to me that the screen refresh process (sending VSYNC, HSYNC, PCLK) has to be running before the panel is configured via SPI. If the configuration via SPI takes place before the screen refresh timing has started ,it would be resetted due to the not available VSYNC.
Could be the other drivers you tested handle this different than the Lovyan ?  Just guessing ....

I hope to have a 86S version panel on hand soon otherwise troubleshooting is not reasonable.

**Reply by moinmoin-sh** [reply]
may be worth to test with  

 `cfg.vsync_polarity = 0`  ?

**Reply by strange-v** [reply]
I've already tried with different polarities and timings (but guessing timings is probably not the best approach).
The panel works with Arduino_GFX with the same polarities and timings. Here is the code:
```cpp
bus = new Arduino_SWSPI(
    GFX_NOT_DEFINED, /* DC */
    TFT_CS,          /* CS */
    TFT_SCLK,        /* SCK or SCLK */
    TFT_MOSI,        /* MOSI */
    GFX_NOT_DEFINED  /* MISO */
);
panel = new Arduino_ESP32RGBPanel(
    TFT_DE,    /* DE */
    TFT_VSYNC, /* VSYNC */
    TFT_HSYNC, /* HSYNC */
    TFT_PCLK,  /* PCLK */
    TFT_D11,   /* R0 */
    TFT_D12,   /* R1 */
    TFT_D13,   /* R2 */
    TFT_D14,   /* R3 */
    TFT_D15,   /* R4 */
    TFT_D5,    /* G0 */
    TFT_D6,    /* G1 */
    TFT_D7,    /* G2 */
    TFT_D8,    /* G3 */
    TFT_D9,    /* G4 */
    TFT_D10,   /* G5 */
    TFT_D0,    /* B0 */
    TFT_D1,    /* B1 */
    TFT_D2,    /* B2 */
    TFT_D3,    /* B3 */
    TFT_D4,    /* B4 */
    1,         /* hsync_polarity */
    8,         /* hsync_front_porch */
    10,        /* hsync_pulse_width */
    50,        /* hsync_back_porch */
    1,         /* vsync_polarity */
    8,         /* vsync_front_porch */
    10,        /* vsync_pulse_width */
    20         /* vsync_back_porch */
);
gfx = new Arduino_RGB_Display(
    TFT_WIDTH,  /* width */
    TFT_HEIGHT, /* height */
    panel,
    0,    /* rotation */
    true, /* auto_flush (IPS, in makerfabs project it is IPS) */
    bus,
    41, /* RST */
    gc9503v_type1_init_operations,
    sizeof(gc9503v_type1_init_operations));
```
I agree that this is something related to reset. As you can see above, I initialized the display with RST pin (41), otherways it doesn't work.
While playing with LovyanGFX I tried with and without reset.

**Reply by moinmoin-sh** [reply]
I agree that guessing is not the best approach and for that reason I'll stop for now and come back once I'm able to test by my own.
Please post here if you have any success.

**Reply by moinmoin-sh** [reply]
Can't stop guessing ...

2 things 
- reset has to be configured to pin 41
- the reset timing in Lovyan's driver is too short for the used reset circuit in 86S. The following change could be made to` PanelDevice.cpp` to increase the reset timing to 50 ms low (as it's done in Wireless sample code)
```C
  void Panel_Device::reset(void)
  {
    auto pin = _cfg.pin_rst;
    if (pin < 0) return;
    gpio_hi(pin);
    delay(100);  //old : 64
    gpio_lo(pin);
    delay(50);   //old: 4
    gpio_hi(pin);
    delay(100);  //old: 64
  }

```
untested, but would make sense to me taking into account what was reported above.

update:
I checked the Arduino_GFX also and found there a reset low time of 120 ms :
https://github.com/moononournation/Arduino_GFX/blob/master/src/display/Arduino_RGB_Display.cpp#L36

an additional indication that the 4ms in the Lovyan driver is too short

**Reply by strange-v** [reply]
Unfortunately, it doesn't work (despite sounding logical). I've tried low time 50, 120, 500 and got the same black screen.

**Reply by moinmoin-sh** [reply]
Just to make sure, did you set       `cfg.pin_rst = GPIO_NUM_41;   ?
`

**Reply by strange-v** [reply]
Yes
![image](https://github.com/lovyan03/LovyanGFX/assets/3875529/53d6f50d-42e5-4210-a66e-d384815c2454)

**Reply by moinmoin-sh** [reply]
Can you make sure that the modified code in the Lovyan library is running, eg. by inserting a serial output.
Sorry for bothering, but that's what I would do if I had the HW to confirm myself.

**Reply by strange-v** [reply]
@moinmoin-sh, yes, I already tried it (this is also my go-to way of testing), and it does print from the `Panel_Device::reset`.
I switched to Arduino_GFX for now, but it's not a problem for me to test something that may fix the issue.

Ideally, I'd like to compare the performance of LovyanGFX and Arduino_GFX  and choose the fastest.

**Reply by moinmoin-sh** [reply]
OK, thanks for testing. I was pretty much convinced that this would fix the issue. May be there is something more ... ?

**Reply by tobozo** [reply]
what if the reset is done manually, before tft.init() and the pin_rst stays unassigned? 

    cfg.pin_rst = -1;

**Reply by moinmoin-sh** [reply]
@strange-v can you try that suggestion?

**Reply by strange-v** [reply]
Finally!
![image](https://github.com/lovyan03/LovyanGFX/assets/3875529/0db4803d-c2c7-4cd4-bd19-d5c73015b83f)

Thank you @tobozo and @moinmoin-sh.

**Reply by moinmoin-sh** [reply]
Nice that it works, but would be much nicer to understand why ...
(what is different between the modified library reset function and the manual reset function ?)

**Reply by tobozo** [reply]
could it be because VSYNC and RST use different directions?

**Reply by moinmoin-sh** [reply]
I don't think so. VSYNC and RST are both outputs.

It's interesting why the same function outside the library works as expected, but inside not.

**Reply by moinmoin-sh** [reply]
May be it would help to change the order here:
```C
  bool Panel_Device::init(bool use_reset)
  {
    init_cs();
    _bus->init();
    init_rst();
    if (_light)
    {
      _light->init(0);
    }
    if (use_reset)
    {
      reset();
    }
    return true;
  }
```
put `init_rst()` and `reset()` before `_bus->init()` because reset and VSYNC are using the same pins and reset is used before the bus is.

**Reply by strange-v** [reply]
I'll be able to test it later on today.

**Reply by tobozo** [reply]
could it be overriden form a `Panel_GCxxxx.cpp` file instead of modifying the general behaviour of the Panel_Device class?

see [Panel_IT8951.cpp::init(bool use_reset)](https://github.com/lovyan03/LovyanGFX/blob/master/src/lgfx/v1/panel/Panel_IT8951.cpp#L128)

**Reply by moinmoin-sh** [reply]
After some thinking I come to the following conclusion:

the bus is initialized first (includes the pin41 = VSYNC)
then the reset is initialized (redefines pin41 = RST)
then the reset sequence runs, leaving pin41 in HIGH state
as a consequence the pin41 doesn't belong to the bus anymore -> we don't get VSYNC -> screen stays black

Changing / overriding the Reset function wouldn't change anything, it needs to be in the right order (Reset before BUS)

just my 5 cent

**Reply by strange-v** [reply]
Modifying `Panel_Device::init` to reset before the bus initialization fixes the issue.
```cpp
bool Panel_Device::init(bool use_reset)
{
  init_cs();
  init_rst();
  if (use_reset)
  {
    reset();
  }
  _bus->init();
  if (_light)
  {
    _light->init(0);
  }
  return true;
}
```
The reset timings can be left as they were before (64, 4, 64).

**Reply by moinmoin-sh** [reply]
Thanks for testing! That confirms my theory very well I think.
Nevertheless the reset works with the short low time, I would prefer to increase it to 32 ((64, 32, 64) to stay in the nuber schema) to ensure some safety margin. 32 msec during init doesn't hurt and should be good enough for the future.

**Reply by tobozo** [reply]
> Changing / overriding the Reset function wouldn't change anything,

The suggestion is to override the `init()` method so that other LGFX panels aren't affected by a change that only addresses problems with `GC9503V` display.

Ideally the custom `init()` should compare RST and VSYNC pin values and only use a custom reset when both values match.

**Reply by lovyan03** [reply]
I have now updated the develop branch with changes regarding the panel reset sequence.
Perhaps it will work even if you are using both the reset pin and VSYNC, but try it and see.

Then `lgfx::Panel_GC9503` is in `src/lgfx/platforms/esp32s3/Panel_RGB.cpp` .
If you want to make changes to the contents of init, make changes to the init method of this class.

**Reply by strange-v** [reply]
@lovyan03 the code from dev doesn't work, unfortunately.
To make it work, I have to change `Panel_Device::init` like this:
```cpp
bool Panel_Device::init(bool use_reset)
{
  init_rst();
  init_cs();
  if (_light)
  {
    _light->init(0);
  }
  if (use_reset)
  {
    rst_control(false);
    delay(8);
  }
  rst_control(true);
  delay(16); // doesn't work with a value less than 9
  _bus->init();
  return true;
}
```

UPD: To make those changes in Panel_GC9503::init we'll have to duplicate the code from Panel_Device::init, as it uses this method.

**Reply by lovyan03** [reply]
@strange-v 
I have made adjustments again and updated the develop branch.

It is very important, but I believe there are panels that will not work if I adopt your suggested sequence.
The reset must be completed after the bus initialization is completed.
The order in which bus initialization is performed after the reset is completed cannot be adopted.

**Reply by strange-v** [reply]
It works with the latest changes!

**Reply by lovyan03** [reply]
Did you solve the problem?
If you really want to do a special reset, you can do the following

```
#include <LovyanGFX.hpp>
#include <lgfx/v1/platforms/esp32s3/Bus_RGB.hpp>
#include <lgfx/v1/platforms/esp32s3/Panel_RGB.hpp>

class LGFX : public lgfx::LGFX_Device
{
  lgfx::Panel_GC9503 _panel_instance;
  lgfx::Bus_RGB       _bus_instance;
public:
  const gpio_num_t PIN_RST = GPIO_NUM_41;

  bool init_impl(bool use_reset, bool use_clear) override
  {
    lgfx::pinMode(PIN_RST, lgfx::pin_mode_t::output);
    lgfx::gpio_hi(PIN_RST);
    lgfx::delay(64);
    lgfx::gpio_lo(PIN_RST);
    lgfx::delay(8);
    lgfx::gpio_hi(PIN_RST);
    lgfx::delay(64);
    return LGFX_Device::init_impl(use_reset, use_clear);
  }
```

**Reply by strange-v** [reply]
@lovyan03 the latest changes in dev resolved the issue with WT32S3-86S. But thanks for the example of a custom reset.

**Reply by alpha815** [reply]
Hi guys, Sorry i am bit late to part. I recently got this display and it is working with Arduino_GFX library. But I would love to get a sample code for LovyanGFX. If you have it sorted out, can you provide a "hello world" sample for this display?

**Reply by strange-v** [reply]
@alpha815, I don't have a standalone example, but you can use [this one](https://github.com/strange-v/WT-86-32-3ZW1_LVGL) (it's for a different screen) with this [LGFX.hpp](https://github.com/strange-v/ha_deck/blob/main/components/hd_device_wt32s3_86s/LGFX.hpp) file.
